Controller: use EasyMesh messages for access control

CoralMalachi opened the pull request:
The client steering task should use the appropriate EasyMesh CMDU.

Change client steering task to use Client Association Control request
messages instead of vendor-specific messages.

Fixes #316, #317 

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf:tlvClientAssociationControlRequest auto-generated files update

Preparative commit.
Add "tlvf/wfa_map/tlvClientAssociationControlRequest.yaml" to
tlv_conf.yaml file and add its auto generated files as well.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
controller:use EasyMesh messages for access control

The client steering task should use the appropriate EasyMesh CMDU.
Therefore the following changes were made:

Change client steering task to use Client Association Control request
messages instead of vendor-specific messages.

Replace intel proprietary message client control allow/disallow
request, by a dedicated 1905 CMDU.
Add as well vendor-specific tlv which contains some data which does not
exist on 1905 CMDU, to keep backward compatibility.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

arnout commented on controller/src/beerocks/master/tasks/client_steering_task.cpp:
```diff
@@ -92,20 +93,36 @@ void client_steering_task::steer_sta()
+    association_control_request_tlv->alloc_sta_list();
+    auto sta_list         = association_control_request_tlv->sta_list(0);
+    std::get<1>(sta_list) = network_utils::mac_from_string(sta_mac);
+    association_control_request_tlv->bssid_to_block_client() =
```
Please keep the same order as in the TLV, so first set bssid, then association control, then validity period, and only at the end allocate the station list.

arnout commented on controller/src/beerocks/master/tasks/client_steering_task.cpp:
```diff
@@ -92,20 +93,36 @@ void client_steering_task::steer_sta()
+        network_utils::mac_from_string(current_ap_mac);
+    association_control_request_tlv->validity_period_sec() = disassoc_timer;
+    association_control_request_tlv->association_control() =
+        wfa_map::tlvClientAssociationControlRequest::BLOCK;
```
This  should be an UNBLOCK. The blocking is done below (cACTION_CONTROL_CLIENT_DISALLOW_REQUEST). That one should also be updated.

arnout reviewed (changes requested):
Typo in the second commit message: Associan -> Association.

As advised by @morantr, we should only merge the PR when it is complete, i.e.:

- controller uses the message;
- agent handles the message;
- dummy bwl reports it;
- a test is added for this flow, by issuing a bml command that triggers it, and verifying the bwl log message on the agent.
- VS handling is removed from agent;
- action is removed.

Commit:
common: Add vendor-specific tlv

Preparative commit.

The following commit changes client steering task to use Client
Association Control request messages instead of vendor-specific messages.

There is information that is generated by client steering task
that is not included in the standard Client Association Control Request
message.

For that reason, a new vendor-specific TLV is defined to include this
information, which will be sent as part of 1905
message:  "tlvVsClientAssociationControlRequest".

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
agent: add client association request handler

The agent handles the Client Association Request.

In case of blocking request, the agent creates a vendor-specific message
of type cACTION_APMANAGER_CLIENT_DISALLOW_REQUEST, else
cACTION_APMANAGER_CLIENT_ALLOW_REQUE, and send it to the
ap_manager_socket.

Also, dummy bwl is updated to reports it, by logging a suitable
message.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
agent: Remove beerocks access control handlers

Remove intel proprietary client control allow/disallow
request message handler on son_slave_thread.
This handler has no use anymore.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
common: tlvf: Remove intel propriety control client allow/disallow messages

Remove yaml and auto-generated files of intel propriety
allow/disallow control messages. These messages have no longer have
use because replaced with easymesh message.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tools: docker: update client steering dummy flow test

Add this flow to test_client_steering_dummy test, which triggers the
sending of a Client Steering Request message using beerocks_cli and
dummy BWL.

Therefore the following changes were made:
send a client association control request to the chosen BSSID to steer
the client (BLOCK).

Send client association control request to all other BDDID (BLOCK)

Add checks to validate the client allow and disallow requests were received
by the correct repeater and the entire flow continued as expected.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_1905_vs.yaml:
```diff
@@ -35,3 +35,9 @@ tlvVsClientAssociationEvent:
+
+tlvVsClientAssociationControlRequest:
+  _type: class
+  reject_sta: uint8_t
```
reject_sta is only used by the wav and dwpal BWL implementations. However, upstream hostapd doesn't have anything like that, and I don't really understand what its purpose could be.

That said, the Multi-AP specification does say what the reject reason should be:

>  it shall set the Status Code in the Authentication frame or (Re-)Association Response frame to a value that
> does not indicate capabilities mismatch or negotiation failure, and should set the Status Code to value 33 (denied due to
> insufficient bandwidth) or 34 (denied due to poor channel conditions).

So the 33 *does* make sense. However:

- it should always be 33, not 0;
- it's the agent's responsibility to take care of this.

So I think we should remove reject_sta here, and instead handle it inside BWL itself.

Of course, that's not for this PR... -> #430

CoralMalachi replied:
Ok thanks 👍

CoralMalachi replied:
Why it should always be 33? 
In case of denied due to poor channel conditions, it should not be 34? @arnout

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_1905_vs.yaml:
```diff
@@ -35,3 +35,9 @@ tlvVsClientAssociationEvent:
+tlvVsClientAssociationControlRequest:
+  _type: class
+  reject_sta: uint8_t
+  ipv4: beerocks::net::sIpv4Addr 
```
ipv4 is not used for anything except a log message, so I don't think we need to add it. nitpick, can be solved later.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4780,6 +4733,44 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
+        message_com::add_vs_tlv<beerocks_message::tlvVsClientAssociationControlRequest>(cmdu_tx);
+    if (!vs_tlv) {
+        LOG(ERROR) << "add_vs_tlv tlvVsClientAssociationControlRequest failed";
+        return false;
```
A missing vs_tlv should *not* be fatal!

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4780,6 +4733,44 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
+            LOG(ERROR) << "Failed building ACTION_APMANAGER_CLIENT_ALLOW_REQUEST message!";
+            return false;
+        }
+        request_out->ipv4() = vs_tlv->ipv4();
```
Note that vs_tlv may be nullptr.

CoralMalachi replied:
But I'm checking for this here :
`
if (!vs_tlv) {
        LOG(ERROR) << "add_vs_tlv tlvVsClientAssociationControlRequest failed";
        return false;
    }
`
You mean I should not return false and use this vs_tlv only if it valid ? @arnout

arnout replied:
> But I'm checking for this here :
> `if (!vs_tlv) { LOG(ERROR) << "add_vs_tlv tlvVsClientAssociationControlRequest failed"; return false; }`
> You mean I should not return false and use this vs_tlv only if it valid ? @arnout

Yes, that's what I meant with "a missing TLV should not be fatal".

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4780,6 +4733,44 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
+            LOG(ERROR) << "Failed building ACTION_APMANAGER_CLIENT_DISALLOW_REQUEST message!";
+            return false;
+        }
+        request_out->reject_sta() = vs_tlv->reject_sta();
```
Note that vs_tlv may be nullptr.

arnout commented on controller/src/beerocks/master/son_actions.cpp:
```diff
@@ -154,14 +156,27 @@ void son_actions::unblock_sta(db &database, ieee1905_1::CmduMessageTx &cmdu_tx,
-        if (request == nullptr) {
-            LOG(ERROR) << "Failed building ACTION_CONTROL_CLIENT_ALLOW_REQUEST message!";
+        if (!cmdu_tx.create(0,
+                            ieee1905_1::eMessageType::CLIENT_ASSOCIATION_CONTROL_REQUEST_MESSAGE)) {
```
There are three places in the controller where very similar code is used to build the Client Association Control Request message. It might be worth moving that to a single function. Unfortunately, we don't have a good place for it...

(nitpick)

arnout commented on controller/src/beerocks/master/son_actions.cpp:
```diff
@@ -154,14 +156,27 @@ void son_actions::unblock_sta(db &database, ieee1905_1::CmduMessageTx &cmdu_tx,
+            network_utils::mac_from_string(current_ap_mac);
+        association_control_request_tlv->association_control() =
+            wfa_map::tlvClientAssociationControlRequest::UNBLOCK;
+        //TODO: Get real validity_period_sec
```
I think this is only valid for blocking, not for unblocking, so it could be set to 0. But I don't find the relevant part of the MultiAP spec.

CoralMalachi replied:
I think you are right :
`The Validity Period field in a Client Association Control Request message with Association Control field set to 0x01 (Client Unblocking) is ignored`  
So I'll change its value to zero.

arnout commented on controller/src/beerocks/master/tasks/client_steering_task.cpp:
```diff
@@ -148,24 +172,47 @@ void client_steering_task::steer_sta()
+            network_utils::mac_from_string(current_ap_mac);
+        association_control_block_request_tlv->association_control() =
+            wfa_map::tlvClientAssociationControlRequest::BLOCK;
+        association_control_block_request_tlv->validity_period_sec() = disassoc_timer;
```
I don't think this is correct... First of all, of course, disassoc_timer is expressed in TU while validity period is expressed in seconds, so that's a factor 1000 different. But also, the block should definitely still be in place immediately after the disassociation took place. So a more reasonable validity period, I think, would be the timeout used by the client steering task, i.e. steering_wait_time_ms/1000. In fact, that would remove the need to do unblocking in the FINALIZE state.

But again, that's for a different day. Could you make an issue for it?

CoralMalachi replied:
Yes 👍 @arnout

CoralMalachi replied:
#432

arnout reviewed (approved):
The commit "agent: Replace control client request to easymesh" is actually for the controller. And there's a separate commit "controller:use EasyMesh messages for access control" that *also* changes the controller. It's not so clear based on what it is split like that.

Also, it would be better to have the controller change come *after* the agent change. Indeed, access control no longer works until the agent is updated, while vice versa it didn't work anyway with EasyMesh messages so adding support for them can't break anything. But that's a small thing.

Except for these remarks, the splitting and ordering of commits and the commit messages are excellent.

The rest of the review is mostly nitpicks. Even not failing on a missing VS message is something that can be fixed later, so I will approve already.
